File: python_examples/demos/web_search_multimodal.py
#!/usr/bin/env python3
"""Web search multimodal demo - find and download video clips."""

import asyncio
import os, re, sys
from pathlib import Path
from dotenv import load_dotenv
from mix_python_sdk import Mix
from mix_python_sdk.helpers import send_with_callbacks
from mix_python_sdk.tool_models import MediaShowcaseParams, CoreToolName

# Add parent directory to path for utils import
sys.path.insert(0, str(Path(__file__).parent.parent))
from utils import upload_to_storage


async def main():
    load_dotenv()

    async with Mix(server_url=os.getenv("MIX_SERVER_URL")) as mix:
        mix.preferences.update_preferences(
            preferred_provider="anthropic",
            main_agent_model="claude-sonnet-4-5",
        )

        session = mix.sessions.create(title="Web search multimodal demo")

        downloaded_files = []

        def handle_tool(t):
            if t.name != CoreToolName.SHOW_MEDIA or not t.input:
                return
            for o in MediaShowcaseParams.model_validate_json(t.input).outputs:
                print(f"\n{o.title}\n   {o.description or ''}\n   {o.path or ''}\n")
                if o.path:
                    downloaded_files.append(o.path)

        await send_with_callbacks(
            mix,
            session_id=session.id,
            message="First, find the top 3 karpathy LLM videos, then find the most important 10 second section from each video. After that, download the sections and show it.",
            on_tool=handle_tool,
            on_content=lambda text: print(
                re.sub(r"(\.)([A-Z])", r"\1\n\2", text), end="", flush=True
            ),
            on_error=lambda error: print(f"\n❌ {error}"),
        )

        # Upload downloaded files to the storage(local/supabase)
        # if downloaded_files:
        #     for file_url in downloaded_files:
        #         await upload_to_storage(mix, session.id, file_url)


if __name__ == "__main__":
    asyncio.run(main())

File: typescript_examples/demos/web_search_multimodal.ts
#!/usr/bin/env tsx
/**
 * Web search multimodal demo - find and download video clips.
 */

import { config } from "dotenv";
import { Mix } from "mix-typescript-sdk";
import { sendWithCallbacks } from "../helpers.js";
import { uploadToStorage } from "../utils.js";

config();

async function main() {
  const mix = new Mix({
    serverURL: process.env.MIX_SERVER_URL,
  });

  // Update preferences
  await mix.preferences.update({
    preferredProvider: "anthropic",
    mainAgentModel: "claude-sonnet-4-5",
  });

  // Create session
  const session = await mix.sessions.create({
    title: "Web search multimodal demo",
  });

  const downloadedFiles: string[] = [];

  // Handle tool events
  function handleTool(tool: any) {
    if (tool.name !== "ShowMedia" || !tool.input) return;

    try {
      const input = typeof tool.input === "string" ? JSON.parse(tool.input) : tool.input;
      const outputs = input.outputs || [];

      for (const output of outputs) {
        console.log(`\n${output.title}`);
        if (output.description) console.log(`   ${output.description}`);
        if (output.path) {
          console.log(`   ${output.path}`);
          downloadedFiles.push(output.path);
        }
        console.log();
      }
    } catch (e) {
      // Silently ignore parse errors
    }
  }

  // Search and create video clips
  await sendWithCallbacks(
    mix,
    session.id,
    "First, find the top 3 karpathy LLM videos, then find the most important 10 second section from each video. After that, download the sections and show it.",
    {
      onTool: handleTool,
      onContent: (text) => {
        // Add line breaks after periods before capital letters for readability
        const formatted = text.replace(/(\.)([A-Z])/g, "$1\n$2");
        process.stdout.write(formatted);
      },
      onError: (error) => console.log(`\n❌ ${error}`),
    }
  );

  // Upload downloaded files to storage (local/supabase)
  // Uncomment to enable uploading to storage
  // if (downloadedFiles.length > 0) {
  //   for (const fileUrl of downloadedFiles) {
  //     await uploadToStorage(mix, session.id, fileUrl);
  //   }
  // }
}

main()
  .catch(console.error)
  .finally(() => process.exit(0));
